README.md
# KRTPluginA
![Donate](PaypalDonate.png)

[Donate](https://www.paypal.com/donate/?hosted_button_id=Y94P8UE7PYCJY)

  * Escape the `[dev-bot]` spam of a watch by a filtered removal of `[dev-bot]`

Current development on [Linux ARM](https://drive.google.com/drive/folders/11k81iJwAW_xP3fNGO9ZmNh-XwS2DexbF) Chromebook. 

Version: `RACK_VERSION.WORKING_MACHINES_COUNT.FIX_INCREMENT` supplied in tags. Any commit not tagged is not a version for compile, although it may work it could also be quite dangerous. The index is numbered so that `WORKING_MACHINES_COUNT` makes sense given the development ordering of the index.

Current public version: [here](https://library.vcvrack.com/KRTPluginA) archives for older versions of the demos maybe available [here](https://github.com/jackokring/KRTPluginADemo).

Compiled with: `gcc (Debian 10.2.1-6) 10.2.1 20210110` or later. Any issues due to edge case optimizations of `float` may be corrected if they are problems.

Less technical: [Captain's log](https://jackokring.github.io/KRTPluginADemo/rocket.html) for less detail and more of an overview.

## Other Plugins (Ideas and Developments)
[KRTPluginJ](https://github.com/jackokring/jqt-chromebook-arm/tree/libj-wrapper) might be an interesting idea. It must be GPL3 however due to licence restrictions.

## Index - Version - Demo

The version number indicates the API build compatibility and feature use. So 1 is VCVRack V1 API. These will be changed as higher API number features are used and adapted to.

1. [A](#a) - 2 - [DA](https://github.com/jackokring/KRTPluginADemo/blob/master/DA.vcv)
2. [μ](#μ) - 2 - [crisp](https://github.com/jackokring/KRTPluginADemo/blob/master/crisp.vcv)
3. [T](#t) - 2 - [I shat the she riff](https://github.com/jackokring/KRTPluginADemo/blob/master/I%20shat%20the%20she%20riff..vcv)
4. [L](#l) - 2 - [boring](https://github.com/jackokring/KRTPluginADemo/blob/master/boring.vcv)
5. [;D](#d) - 2 - [DA](https://github.com/jackokring/KRTPluginADemo/blob/master/DA.vcv)
6. [R](#r) - 2 - No demo required
7. [Ω](#Ω) - 2 - [Omicron](https://github.com/jackokring/KRTPluginADemo/blob/master/Omicron.vcv)
8. [V](#v) - 2 - [Omicron](https://github.com/jackokring/KRTPluginADemo/blob/master/Omicron.vcv)
9. [F](#f) - 2 - [morph](https://github.com/jackokring/KRTPluginADemo/blob/master/morph.vcv)
10. [W](#w) - 2 - [W](https://github.com/jackokring/KRTPluginADemo/blob/master/W.vcv)
11. [Y](#y) - 2 - [Y](https://github.com/jackokring/KRTPluginADemo/blob/master/Y.vcv)
12. [O](#o) - 2 - [parabolic](https://github.com/jackokring/KRTPluginADemo/blob/master/parabolic.vcv)
13. [S](#s) - 2 - [SE](https://github.com/jackokring/KRTPluginADemo/blob/master/SE.vcv)
14. [E](#e) - 2 - [SE](https://github.com/jackokring/KRTPluginADemo/blob/master/SE.vcv)
15. [X](#x) - 2 - [wave](https://github.com/jackokring/KRTPluginADemo/blob/master/wave.vcv)
16. [M](#m) - 2 - [M](https://github.com/jackokring/KRTPluginADemo/blob/master/M.vcv)
17. [Q](#q) - 2 - [Q](https://github.com/jackokring/KRTPluginADemo/blob/master/Q.vcv)
18. [C](#c) - 2 - No demo required
19. [U](#u) - 2
20. [B](#b) - 2 - I seems to be adding to patch [B](https://github.com/jackokring/KRTPluginADemo/blob/master/B.vcv)
21. [J](#j) - 2
22. [K](#k) - 2
23. [H](#h) - 2
24. [I](#i) - 2
25. [G](#g) - 2
26. [Z](#z) - 2

[Easy Access Demo Archive](https://github.com/jackokring/KRTPluginADemo/archive/refs/heads/master.zip) includes all the demos above as a **.zip** file for easy downloading and none of that cut, paste, save and rename business.

A
===

*49% Transparent Module - Made with sexspex plastique*

Parameters
---
* Frequency (-12dB) `FRQ`
* Resonance `REZ`
* Frequency Modulation Gain `MOD`
* Relative Plate (-6 dB) `OFF`
* Plate Gain `AM`
* Plate Modulation Gain `MAM`

Ins
---
* -18 dB In (HP on `XP12`) `IN1`
* -18 dB Predictive (LP on `XP12`) `LP1`
* Frequency CV `CV`
* -12 dB In (No Plate) `IN2`
* -6 dB HP In (Plate Ring) `HP1`
* Modulation `MOD`

Outs
---
* -6 dB LP Out `LP1`
* Tracking Out `CV`
* -12/-18 dB Out `LP12`
* -6 dB HP Out `HP1`
* Plate Out `AM`
* -12/-18 dB Inverse Extended Filter Out `XP12`

A generic filter with some nice fun additions for a characteristic sound. How much fun can you have with 4 poles and some other DSP functions? Yes, it is by design digital, although it might be possible to make an analogue version of most if not all of the implied circuit.

The filter poles are ZDF (Zero Delay Filter) arithmetic. The `LP1` predictive input creates a
ninth order estimate one sample into the future. This has the effect of a bit of phasing and given that it is not equally fed into the ring modulation too, makes it spectrally different. The 45 degree lead lag of the ring makes for a ninety on the AM, for even harmonic self AM power.

If some controls appear not to operate use `IN1` and `LP12` out, with `CV` for tracking, and then all the controls work. Some specific inserts are pre or post certain controls. The plate AM tracking stacks on the filter corner tracking.

μ
===

*The Gold One - Summation generative additive synthesis*

**SIGNATURE**

Parameters
---
* Master Gain `dB`
* Slew Limit `Hz`
* Halflife `λ`
* Gain 1 `G1`
* Gain 2 `G2`
* Gain 3 `G3`

Ins
---
* 1V/6dB `CVdB`
* 1V/Oct `CVHz`
* 1V/halflife  `CVλ`
* In 1 `IN1`
* In 2 `IN2`
* In 3 `IN3`

Outs
---
* Differential `Δ`
* Differential `ΔΔ`
* Differential `ΔΔΔ`
* Integral Linear `Σ`
* Integral Pole `1/Σ`
* Integral Log `LNΣ`

Some weird calculus module. It really depends what it can be made to do. It might become unstable if you feedback connections. It might not as well. The `λ` control only affects the last 5 outputs, as the 3 integral representations should converge to the same value and `λ` affects the effective sample timing offset next to a potential singular solution.

All the estimators are predictive one sample into the future (with `λ` kind of being a sophisticated constant of integration), a pre slew filter is set as a 6 dB/Oct zero delay filter. This filter may work opposite to how you might expect for some outputs.

So the inputs `IN1`, `IN3` and `IN3` are controlled by **centred dB scaling** gainuverters for gain, amplified by an exponential `dB`, slew limited and then processed for calculus estimates (9th order FIR).

The integrals are further processed by 3 point series accelerations (a tiny effect but interesting) to occur at a sample in the future. `λ` applies an effective modulation of this integral time step from a fraction of a sample to multiple samples. The half-life then is measured in exponential inverse octaves.

I mean why a 9th order FIR, and the number 27 of the sporadic simple groups connection? From 2 comes 2 to the power of all N. From 3 comes all odd numbers? This likely explains the choice of cube root to control the range of output voltage.

Blurb (not essential to understanding the module)
---

He's irresponsible :D https://en.wikipedia.org/wiki/Jacques_Tits ... roll on (Monster Moonshine)[https://en.wikipedia.org/wiki/Monstrous_moonshine] and let's not confuse maths with physics "speculated" on maths. Hope I got all the integration by parts and collection of approximants in limited terms right? (Some 2's, some 3's, some hopeful sporadics on the elliptic). Did I mention the congruants div 3 and the determinants (count 44 (18n+26 (Groupies)[https://en.wikipedia.org/wiki/Sporadic_group])) with Z(p) and alternating groups to a 42 which the circle and ellipse are the alternating or not on the major minor radius?

Initialize the big bang with differential estimates of the future captain, so obviating the definitional requirements at the singularity Captain. Well Gamma Bob, well gamma. A form of precision series truncation crusher, nice! None of that chucking away an abstract precision and getting right in there with truncation of summation generative additive synthesis, for that timbrel union.

But isn't it about the singularity context of integration in a set of limits x, x+y where x is unknown but guesstimated in a relative context Captain? Yes Bob all error in the calculus should be reported as this module is not about error, and others maybe built.

But shouldn't a sine converge to sin(x)? And and truncation of terms provide a controlled deviation from sin(x) for Fourier harmonics? Exactly Boz, the sweet low power law alias summands.

"Proxy off a dirty hole of unknown gain? How dare you? Must fix the incoming to be at least pleasant." -- Captain of the Exponential.

T
===

*The White One - A harmonic pitch down and up delay*

Parameters
---
* Course Tune `NOTE`
* Fine Tune `FINE`

Ins
---
* In `IN`
* Trigger`TRIG`

Outs
---
* High Trigger `HI`
* Out `OUT`

The trigger `TRIG` resets the delay and produces a down shifted `OUT` until the up shift starts causing `HI` to output a rising edge and the delay switches an octave up until the expected retrigger point. The delay plays regular speed until a retrigger. The delay will reset itself if the delay buffer becomes full.

Going down is always a prerequisite before rising to the top! And normal is very overdue. The maximal delay length is divided by polyphony. What's the cache flush GB/s on this thing? I'm thinking 4 bars and a bit or so?

Polyphonic triggers occur at the optimal octave intercept per trigger, and so the outputs become phasic to the triggers. `HI` can off trigger a trigger to sync with the up speed clearing of the precision down beat zero delay. No trigger equals no effect and no delay after the "blip" as there is a trigger need on `TRIG`.

Try `NOTE` at 5 semitones, and alter the trigger rate for a sync chord arpy sound. In this sense other semitone values are better at lower `TRIG` rates. Try some drums on the `HI` out for an interesting timing effect too.

L
===

*A White One - A synchronized delay with pre-trigger end*

Parameters
---
* Pre-trigger Delay `PRE`
* Fine Tune `FINE`

Ins
---
* In `IN`
* Trigger `TRIG`

Outs
---
* Pre-trigger `PRE`
* Out `OUT`

A synchronized delay with a `PRE` control to remove a time dependant on sample latency so as to match timing one synchronization trigger later. So suppose you wish to synchronize to some live play and are prepared to time up one `TRIG` later (say a bar for arguments sake), then you can increase `PRE` to achieve that less than the bar delay with latency compensating the rest of the bar delay which `PRE` removed on `OUT`. The output `PRE` gates high before the bar end to allow down module to trigger other things with the need to be triggered before the bar start.

;D
===

*Another White One - A future estimator VCA with error*

Parameters
---
* Level `dB`
* Level Modulation `CVdB`
* Frequency `FRQ`

Ins
---
* In `IN`
* Level Modulation `CVdB`

Outs
---
* Out `OUT`

Exponential amplitude modulation with a future projection estimate 2 samples into the future for phase critical all pass when used with modules that have a 1 sample delay. The `CVdB` control is a **centred dB scaling** gainuverter.

Also useful for pointing out transients in `IN` as the estimator is excited by the parts of the signal with aliasing potential.

R
===

*The Final White One - A simple mixer split*

Parameters
---
* On Level `ON`
* As Level `AS`

Ins
---
* In `IN`
* Mix Some On `ON`

Outs
---
* Get As `AS`
* Out `OUT`

Mix some `ON` into the `IN` making an `OUT` with control of it `AS` for easy modulation mixing. The controls are **centred dB scaling** gainuverters.

Ω
===

*The 70% Grey One (With A Display) - Clock distribution randomizer*

**SIGNATURE**

Parameters
---
* Bird `BIRD`
* Seed `SEED`
* Randomize `RAND`

Ins
---
* Clock (Polyphony Master) `CLK`
* Reset (Polyphonic OR) `RST`

Outs (Bi-phase on clock given current letter)
---
* `AN`, `BO`, `CP`, `DQ`, `ER`, `FS`, `GT`, `HU`, `IV`, `JW`, `KX`, `LY`, `MZ`

`RST` starts a sequence of gate activations, from the beginning, on the outputs `AN` to `MZ` advanced by the `CLK` trigger. The `SEED` controls the initial symbol and some randomness of the following symbols. `RAND` increases the chaos of randomization from the initial `SEED`. The magic of `BIRD` (the word) applies some modal operator lambda calculus to modify the sequence with the **very** discrete occasional stall in the `MM` Omega bird. This is not the differential calculus of Newton fame, but the one with names such as [Godel](https://en.wikipedia.org/wiki/Kurt_G%C3%B6del) and [Church](https://en.wikipedia.org/wiki/Alonzo_Church).

There is a great book on modal logic called [To Mock a Mocking Bird](https://en.wikipedia.org/wiki/To_Mock_a_Mockingbird), itself a follow up to [Forever Undecided](https://en.wikipedia.org/wiki/Raymond_Smullyan). A lay person introduction to the subject.

Higher polyphony using `CLK` copies the mono output one from the lowest channel to the highest given a `CLK` on a channel. For example when clocked the second channel takes the output of the first channel.

V
===

*The 70% Grey One - Envelope VCA and CV producer*

**DEPRECATED**

Parameters
---
* Base Frequency `Hz`
* Envelope Attack `ATK`
* Envelope Decay `DCY`

Ins
---
* Frequency Modulation `CVHz`
* Amplitude Modulation `CVdB`
* Envelope Time Modulation `CVs`
* 3 of
    * Oscillator In `IN`
    * Up 5 Semitones Trigger `T5`
    * Up 7 Semitones Trigger `T7`

Outs
---
* 3 of
    * Frequency CV to Oscillator`CV`
    * Envelope Shaped Output `OUT`

Takes in triggers on `T5` and `T7` to trigger one of the three vertical envelopes. `IN` is from the oscillator source, and `OUT` is that signal modified by the envelope. `CV` can be tuned dependant on the trigger pair combination. I thought it would help with using Ω. Although it might be quite useful on its own. The two outer `OUT` signals are normalized to the centre `OUT` channel for less cables.

F
===

*The Blue One - Morph filter*

**SIGNATURE**

Parameters
---
* Pole Spread `SPD`
* Pole Skew `SKW`
* Frequency `FRQ`
* Low All High `LAH`
* Invert `INV`
* Drive `DRV`

Ins
---
* Frequency Modulation `FRQ`
* Spread Modulation `SPD`
* Skew Modulation `SKW`
* Low All High Modulation `LAH`
* Invert Modulation `INV`
* Drive Modulation `DRV`
* Input `IN`

Outs
---
* Output `OUT`

A filter with some internal gain processing to handle filter k values into self-oscillation. The four corners set by spread and skew form Linkwitz to Butterworth on spread when skew is full off, and Legendre to Bessel when skew is full on. The concept is stabilized by soft clipping.

W
===

*The Red One - Gate transposer*

Parameters
---
* None

Ins
---
* Input `IN`
* Gates Transpose `1`, `2`, `3`, `4`, `5`, `6`, `7`, `8`, `9`, `10`, `11`

Outs
---
* Output Add `+`
* Output Subtract `-`

To assist in turning patterned gates into CV chromatic sequences. Simple.

Y
===
[Y extra details](Y.md)

*The White Buttons One - Gate sequencer*

Parameters
---
* Tempo `TEMPO`
* Gate Length `G.LEN`
* Four Mode Buttons `PAT`, `SEQ`, `MUTE`, `NOW`
* The Main Buttons (28 for notes and triples)
* Run and Stop `RUN`
* Reset `RST`
* Copy Pattern `CPY` (makes note of active channel for `PAT` paste channel copies)
* Paste Pattern `PST` (only pastes into current channel in `PAT` mode)
* Shuffle Jazz `JAZZ` (three altering micro timings)

Ins
---
* CV `CV`
* Gate `GATE`
* Position Jazz `JAZZ` (later versions)

Outs
---
* Clock `CLK`
* Reset `RST`

A gate pattern sequencer. An obvious layout, but the modes might be complicated.
* `PAT` a pattern edit mode similar to a standard drum machine.
* `SEQ` triggers and selects various sequences. For performance play. Alters assigned pattern on pattern change controls of triples.
* `MUTE` allows channel mutes while maintaining some pattern change control on triples.
* `NOW` puts direct channel gates on the notes (also selects active channel via last pressed channel), while maintaining some pattern change control on triples.

The `CPY` light indicates when the copy source is the same as the paste source. The `PST` light indicates if the copy channel source is the same as the current paste channel.

The `C2` to `B5` keyboard MIDI range on `CV` triggered by `GATE` operate the buttons.
* `C#`, `D#` pair operate function and mode buttons left to right (in pairs).
* `C`, `D`, `E` operate triple buttons.
* `F#`, `G#`, `A#` also operate triple buttons for utility.
* `F`, `G`, `A`, `B` operate quarter buttons.

O
===

*The Beige-ish One - Triple parabolic skew feedback oscillator*

Parameters
---
* Frequency `FRQ` (three)
* Feedback `FBK` (three)

Ins
---
* CV `CV`

Outs
---
* Output `OUT`

An algorithm for sound. Simple to use.

S
===

*A 30% Grey One - MIDI clock signalling to position CV*

Parameters
---
* BPM Estimate `BPM`
* Divide `DIV`

Ins
---
* Clock `CLK`
* Start `STRT`
* Stop `STOP`
* Continue `CONT` (suitable for `Y` module `JAZZ` input)

Outs
---
* Out `OUT`

For interfacing with a MIDI clock provided as an output by the MIDI-CV system module. 64 beats equals 10V and reset back to 0V unipolar signal.

E
===

*Another 30% Grey One - Envelope and VCA*

Parameters
---
* Attack `ATK`
* Release `REL`
* Modulation Level `MOD`

Ins
---
* In `IN`
* Trigger `TRIG`

Outs
---
* Modulation `MOD`
* Out `OUT`

A simple AR envelope with a VCA. The `MOD` control is a **centred dB scaling** gainuverter.

X
===

*Another 30% Grey One - Wave-folder*

Parameters
---
* Fold `FOLD`
* Kind `KIND`
* Mash Noise `MASH`
* Wet Mix `WET`

Ins
---
* In `IN`
* Modulate Fold `MFLD`

Outs
---
* Out `OUT`

An over-sampled wave-folder based on Chebychev polynomials and some extras for a bit of a difference.

M
===

*Another 30% Grey One - Emphasis filter*

Parameters
---
* Low Frequency `LOW`
* Low Gain `LOdB`
* High Frequency `HIGH`
* High Gain `HIdB`

Ins
---
* CV Low `LO`
* CV High `HI`
* In `IN` (send filter input)
* Return `RTN` (return filter input)

Outs
---
* Send `SND` (send filter output)
* Out `OUT` (return filter output)

An equalization filter to decrease the effect of non-linear signal processing on high frequencies throwing them into alias distortion. A return filter removes the emphasis. So the base is amplified and treble reduced on the forward send filter, and the opposite for the return filter. It may make interfacing a "phono" deck easier, but does not amplify the tiny signals from a deck.

It reduces the non-linear shaping of high frequencies in the send return path from injecting harmonics into the alias band above the sampling Nyquist frequency. So good for base harmonic generation. The headroom of VCV rack should support the +/- 50V peak to peak. But you might need a bit of gain reduction before this module.

**WARNING** The default gain settings are good for RIAA equalization, but could be reduced closer to 0 dB for general use. As the return filter is the opposite, that `HIdB` value can be misleading on what the return filter is doing. Under some circumstances it could **destroy tweeter speakers** just as the send filter could wobble the neighborhood walls.

Q
===

*A Green One - Quantum gravimitator*
[The Big Manual?](Q.md)

**SIGNATURE**

Parameters
---
* Omega Frequency `OM`
* Singularity Management `SING`
* Plank `PLK`
* Newton `NTN`
* Mass Ratio `MASS`
* Radial Tangential Angle `ANG`

Ins
---
* Omega Frequency Modulation `OM`
* Singularity Management Modulation `SING`
* Plank Modulation `PLK`
* Newton Modulation `NTN`
* Mass Ratio Modulation `MASS`
* Radial Tangential Angle Modulation `ANG`
* Input `IN`

Outs
---
* Output `OUT`

A physical model simulation. Treats the input as a driven mass radius and calculates the uncertainty of the mass using a quantum and gravity connection.

I could explain by (https://drive.google.com/file/d/0B-avx1luFGRbRTVFOFhSWkluQWM/view?usp=sharing&resourcekey=0-YsVFVTBgEFU-1QAJ1hYz8A)[Mistakes and genius] as yes something are wrong, and something being applied have an error less than Planks constant cubed which is very small as to be immeasurable, yet defined as certain by fixing the constant as a committee committed to, and so ...

Is dark matter the integral over an average on a non-linear 1/r^2 law of an uncertainty in radial potential energy? Is dark energy a tangential version? As distances get smaller the F=kr of quark confinement something that 1/r maps to r becomes a strong force unseen equivalent? Dark strong? A prediction or just a tune from finding sub-harmonic greatest common division of the natural omega frequency of matter?

How to connect your modular to a scanning tunnelling microscope? Contact a physical? Enjoy!


changes.md
2.26.31
===

* `I` phase range moved for easier max (might break patch)

2.26.30
===

* `ConfigSwitch` for some internal state on machines

2.26.29
===

* default bypass?
* Control populator in `plugin.cpp`
* Added envelope externals to `Z`
* Polyphonic `Y` MIDI control

2.26.28
===

* TBC - `Z`
* default bypass on some modules?
* `json_object_set` replaced by `json_object_set_new` memory leak fix
* PayPal donate as GitHub use a bank which ignores Companies House details
* Potential vector optimization
* Fixed `U` independent polyphonic
* Left `V` as some hacky normalized mess
* Fixed `E` modulation bug
* Fixed `E`, `V` envelope bug
* Deprecated `V`
* Fixed `C` missing label crash

2.25.27
===

* complete V2 migration labelling
* improve tool tip documentation

2.25.26
===

* made this file link here ?
* increased SDK V2 (2.25.25) but this file was misdirected
* started adding some V2 labelling to some things
* made `README.md` include version design in index
* decided to follow lowercase convention for labels

## Version 1 SDK Defunct

1.25.25 (actually shows as 1.25.24)
===

* fixed initialization bug regression for some modules added in 1.23.24
* extended control ranges on `G`
* forgot to bump version number so 1.25.24 published is actually 1.25.25
* Final SDK V1 version

1.25.24 (an main build unpublished version number)
===

* clocking on `S` fixed MIDI arm style
* reset on high in `Ω`
* added `I` clock divider
* added `G` compressor

1.23.24
===

* filter zero initializations (all modules)
* increased mash range on `X`

1.23.23
===

* set default `J` mix at 50%
* added `H` phase modulable organ

1.22.22
===

* individual DC filters on `B` (can break patch as now individual filters)
* added `J` all-pass filter/phaser
* added some relevant tags in .json
* added `K` phase modulation oscillator


1.20.21
===

* Made `C` gain normal not SD scaled for 2 ins (via mix) and 1 in split scaling (breaks patch)
* Fixed `C` color on mix label
* Reverted `μ` weighted average drift (fixes patch, mush on 2 of outputs fixed)
* Added `U` sample and hold quantizer
* Added `B` performance router

1.18.21
===

* Fixed `Q` singularity clicks (breaks patch)

1.18.20
===

* Fixed `F` damping as sqrt*sqrt gives correct evaluation of damping and also introduces a sample and hold behavior on high modulation levels in the "full on" corner (slightly breaks patch)
* Added `M` module for base pre-emphasis or RIAA deck preamp with inverse "cutting metal" filter (many VCO can't reach a 20KHz sine high and matching drifts a little)
* Added `Q` module for some effect of a "quantum gravity following" of a source signal
* Added `μ` weighted average drift acceleration algorithm
* Added `C` module for mixing

1.15.20
===

* Fixed `Y` jazz for negative inputs giving better scratching behavior
* Fixed `D` to use oversampling
* Fixed `X` oversampling
* Calmed some of the high sizzle on `μ` (breaks compatibility in a good way) by a 4th order Linkwitz, seems a 3rd differential need a -24 dB on the top for a net -6 dB

1.15.19
===

* Added `X` over-sampled wave folder
* Deleted `O` **.svg** component layer
* Alter `A` estimator to 4th order

1.14.19
===

* Faster `tanf` called `tanpif`
* Color lights on `Y`
* Jazz position input on `Y`
* Added MIDI `S` and ENV/VCA `E`
* Altered `μ` differential calculations to 4th order for less susceptibility to Gibbs oscillations
* Altered `D` estimation calculation to 4th order

1.12.18
===

* `F` k cancels (breaks patch but is functioning as designed)
* `O` works

1.12.16
===

* Added oscillator `O`
* `F` Bessel coefficients from different source (breaks patch)
* `F` frequency fix

1.11.16
===

* Potential `Ω` sharing of display bug fixed (no effect on other function)
* Null terminal in data save `Ω`
* Added in `W` chromatic gate summator
* Added in `Y` gate pattern sequencer

1.9.15
===

* `F` tracks high pass pattern (breaks compatibility)
* `*-rc2` graphics fix colors

1.9.14
===

* Added `F` filter
* Removed `ERR` from `D` and added `FRQ` filter (breaks compatibility)

1.8.13
===

* Fixed version numbering mixup with 1.8.11
* Increased range of halflife on `μ` (breaks compatibility)
* Fixed unit `dBs` (relative log scaled seconds) on `μ` (breaks compatibility)

1.8.11
===

* LPF on `μ` added for slightly less noise on transients.


1.8.10
===

* Fixed percent controls on `Ω` (minor 1 in a million bug)
* Introduced unit `dBs` (relative log scaled seconds) on `V` (breaks compatibility)
* Introduced dB scaling on `A` for `REZ` and `AM` (breaks patch)
* Introduced unit `dBs` (relative log scaled seconds) on `μ` (breaks compatibility)

1.8.9
===

* Reduced scaling on `μ` input `CVdb` (breaks compatibility) 10V/6dB
* Improved documentation
* Website GitHub pages [Captain's log](https://jackokring.github.io/KRTPluginADemo/)

1.6.9
===

* Fixed issues of polyphony
* Fixed time length of buffer `float` pointer in `L`
* Improved documentation
* Improved [demo repository](https://github.com/jackokring/KRTPluginADemo)
* Added cube root output scaling to `u` (breaks compatibility)
* Started change log (this file)


src/G.cpp
#include "plugin.hpp"

struct G : Module {
	enum ParamIds {
		ATK,
		DCY,
		THR,
		RTO,
		CUT,
		Q,
		MIX,
		ENV,
		NUM_PARAMS
	};
	enum InputIds {
		FRQ,
		SCH,
		IN,
		NUM_INPUTS
	};
	enum OutputIds {
		OFRQ,
		OENV,
		OUT,
		NUM_OUTPUTS
	};
	enum LightIds {
		NUM_LIGHTS
	};

	const char *instring[NUM_INPUTS] = {
		"Frequency CV",
		"Side-chain",
		"Audio",
	};

	const char *outstring[NUM_OUTPUTS] = {
		"Frequency CV",
		"Envelope CV",
		"Audio",
	};

	const char *lightstring[NUM_LIGHTS] = {

	};

	void iol(bool lights) {
		for(int i = 0; i < NUM_INPUTS; i++) configInput(i, instring[i]);
		for(int i = 0; i < NUM_OUTPUTS; i++) configOutput(i, outstring[i]);
		if(!lights) return;
		for(int i = 0; i < NUM_LIGHTS; i++) configLight(i, lightstring[i]);
	}

	G() {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
		configParam(ATK, -36.f, -1.f, -9.f, "Attack time", " dBs");
		configParam(DCY, -27.f, 1.f, -6.f, "Decay time", " dBs");
		configParam(THR, -24.f, 6.f, -6.f, "Threshold", " dB");
		configParam(RTO, 8.f, -4.f, 2.f, "Ratio", " 2^N:1");//256 soft-limiter
		//default 4:1 compressing
		configParam(CUT, -8.f, 2.f, 0.f, "Frequency", " Oct");
		//more for base
		configParam(Q, -6.f, 12.f, -6.f, "Resonance", " dBQ");
		configParam(MIX, 0.f, 100.f, 0.f, "Mix gain", " %");
		configParam(ENV, -2.f, 2.f, 0.f, "Envelope amount", " Oct/6dB");
		iol(false);
		for(int i = 0; i < PORT_MAX_CHANNELS; i++) {
			bl[i] = bb[i] = b[i] = last[i] = 0;
		}
	}

	//obtain mapped control value
    float log(float val, float centre) {
        return powf(2.f, val) * centre;
    }

    float qk(float val) {
        float v = log(val, 1.f);
        return 1 / v;//return k from Q
    }

	float f, t, u, k, tf, bl[PORT_MAX_CHANNELS], bb[PORT_MAX_CHANNELS];

    /* 2P
           Ghigh * s^2 + Gband * s + Glow
    H(s) = ------------------------------
                 s^2 + k * s + 1
     */
    //TWO POLE FILTER
	void setFK2(float fc, float q, float fs) {
		//f   = tanf(M_PI * fc / fs);
		f	= tanpif(fc / fs);
		k   = qk(q);
		t   = 1 / (1 + k * f);
		u   = 1 / (1 + t * f * f);
		tf  = t * f;
	}

	float process2(float in, int p) {
		float low = (bl[p] + tf * (bb[p] + f * in)) * u;
		float band = (bb[p] + f * (in - low)) * t;
		float high = in - low - k * band;
		bb[p] = band + f * high;
		bl[p] = low  + f * band;
		return high;//hpf default
	}

	/* 1P H(s) = 1 / (s + fb) */
    //ONE POLE FILTER
	float f1, f2, b[PORT_MAX_CHANNELS], last[PORT_MAX_CHANNELS];

	void setFK1(float fc, float fs) {//fb feedback not k*s denominator
		//f1   = tanf(M_PI * fc / fs);
		f1	 = tanpif(fc / fs);
		f2   = 1 / (1 + f1);
	}

	float process1(float in, int p) {
		float out = (f1 * in + b[p]) * f2;
		b[p] = f1 * (in - out) + out;
		last[p] = out;
		return out;//lpf default
	}

	void process(const ProcessArgs& args) override {
		float fs = args.sampleRate;

		float atk = log(params[ATK].getValue()/3.f, 1.f);
		float dcy = log(params[DCY].getValue()/3.f, 1.f);
		int maxPort = maxPoly(this, NUM_INPUTS, NUM_OUTPUTS);
		float thr = log(params[THR].getValue()/6.f, 5.f);
		float rto = log(-params[RTO].getValue(), 1.f);
		float cut = params[CUT].getValue();
		float rez = params[Q].getValue()/6.f;
		float mix = params[MIX].getValue() * 0.01f;//%
		float env = params[ENV].getValue();//oct/6dB
		float makeUp = 10.f / ((10.f - thr) * rto + thr);

#pragma GCC ivdep
		for(int p = 0; p < maxPort; p++) {
			float frq = inputs[FRQ].getPolyVoltage(p) + cut;
			float in = inputs[IN].getPolyVoltage(p);
			float sch = inputs[SCH].isConnected() ?
				inputs[SCH].getPolyVoltage(p) : in;

			sch = abs(sch);//absolute value
			float tau;
			//f = 1/2pi-tau
			if(sch > last[p]) {
				//attack
				tau = atk;
			} else {
				//decay
				tau = dcy;
			}
			float tasf = 0.5f / (M_PI * tau);//time as frequency
			tasf = clamp(tasf, 0.f, fs * 0.5f);
			setFK1(tasf, fs);
			sch = process1(sch, p);//envelope follow
			float use = 1.f;
			if(sch > thr) {
				//alter use ratio
				float over = sch - thr;
				over *= rto;//compress
				use = (over + thr) / sch;
			}
			in *= use;//apply ratio
			float cv = frq + env * -log2f(use);
			frq = log(cv, dsp::FREQ_C4);//get corner
			frq = clamp(frq, 0.f, fs * 0.5f);
			setFK2(frq, rez, fs);
			in = process2(in, p);//cut and rez
			//make up gain
			in = (1.f - mix) * in + mix * in * makeUp;
			// OUTS
			outputs[OFRQ].setVoltage(cv, p);
			outputs[OENV].setVoltage(sch, p);
			outputs[OUT].setVoltage(in, p);
		}
	}
};

//geometry edit
#define HP 5
#define LANES 2
#define RUNGS 7

struct GWidget : ModuleWidget {
	GWidget(G* module) {
		setModule(module);
		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/G.svg")));

		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

		addParam(createParamCentered<RoundBlackKnob>(loc(1, 1), module, G::ATK));
		addParam(createParamCentered<RoundBlackKnob>(loc(2, 1), module, G::DCY));
		addParam(createParamCentered<RoundBlackKnob>(loc(1, 2), module, G::THR));
		addParam(createParamCentered<RoundBlackKnob>(loc(2, 2), module, G::RTO));
		addParam(createParamCentered<RoundBlackKnob>(loc(1, 3), module, G::CUT));
		addParam(createParamCentered<RoundBlackKnob>(loc(2, 3), module, G::Q));
		addParam(createParamCentered<RoundBlackKnob>(loc(1, 4), module, G::MIX));
		addParam(createParamCentered<RoundBlackKnob>(loc(2, 4), module, G::ENV));

		addInput(createInputCentered<PJ301MPort>(loc(1, 5), module, G::FRQ));
		addInput(createInputCentered<PJ301MPort>(loc(1, 6), module, G::SCH));
		addInput(createInputCentered<PJ301MPort>(loc(1, 7), module, G::IN));

		addOutput(createOutputCentered<PJ301MPort>(loc(2, 5), module, G::OFRQ));
		addOutput(createOutputCentered<PJ301MPort>(loc(2, 6), module, G::OENV));
		addOutput(createOutputCentered<PJ301MPort>(loc(2, 7), module, G::OUT));
	}
};


Model* modelG = createModel<G, GWidget>("G");
